server/src/main/java/org/apache/druid/client/CachingClusteredClient.java

```diff
@@ -462,6 +462,10 @@ private Set<SegmentServerSelector> computeSegmentsToQuery(
               chunk.getChunkNumber()
           );
           segments.add(new SegmentServerSelector(server, segment));
+          if (server.isEmpty()) {
+            // fail the query or alternatively set some headers
+            throw new RuntimeException("Query failed, incomplete data");
```

Comment:
Maybe throw a `QueryException` or a `QueryInterruptedException` with some more info, such as the segment/interval which was found to be unavailable. 

I also think that instead of failing fast, it would be nice if we could identify intervals which are incomplete, then log them, then decide if we want to fail or not. It would be helpful for users to know all the intervals which are incomplete in one go, rather than firing the query and getting it failed multiple times.

Reply:
Throwing `QueryException` with details of unavailable segment ids when the config is set to `fail`.
Also, logging a warning message with the details.